### PR TITLE
time: panic in release when timer polled after completing

### PR DESCRIPTION
**Pull request by davidpdrsn**
While this `debug_assert` could only be [triggered through UB][post]
panicking in release mode and including a more helpful error message is
nice.

Fixes https://github.com/tokio-rs/tokio/issues/3675

[post]: https://fasterthanli.me/articles/pin-and-suffering

### CODE REVIEW

File: tokio/src/time/driver/entry.rs [hunk 1]
```diff
@@ -164,20 +164,23 @@ impl StateCell {
...
         loop {
-            debug_assert!(cur_state < STATE_MIN_VALUE);
+            // Check to see if the timer is already fired. Since firing the
+            // timer can only happen with the driver lock held, we know we
+            // shouldn't be able to "miss" a transition to a fired state, even
+            // with relaxed ordering.
+            //
+            // In part also to improve the error message for things like
+            // https://github.com/tokio-rs/tokio/issues/3675
+            assert!(cur_state < STATE_MIN_VALUE, "timer polled after completing");
```
Comment (Darksonn) [inline tokio/src/time/driver/entry.rs:177]: Is this the right error message here? Can it actually happen if it is polled after completion too, or does this require moving a pinned `Sleep`?

**Comment by Darksonn** [comment]
Ping

**Comment by Darksonn** [comment]
Any updates on this?

**Comment by davidpdrsn** [comment]
Ah damnit sorry. I totally forgot about this 😞 Sorry about this.

I'll just close this and let someone else fix it. That'll probably be quicker.

**Pull request closed without merging**